### ISSUE TITLE
DNS server does not properly recurse

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5524]
|<img alt="greezybacon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| greezybacon reported|
|-|-|
|Trac ID|trac#5524|
|Type|defect|
|Created|2012-03-12 03:23:08Z|
|Branch|https://github.com/twisted/twisted/tree/dns-recursion-5524-2|

The DNS server recurses before checking the hosts file. I added the following to a hosts.txt file:

```
127.0.0.1 bing.com
127.0.0.1 bogus.domain.tld
```

The server was kicked off to serve from the hosts file

```
sudo bin/twistd -n dns --hosts-file=hosts.txt -i 127.0.0.1
```

Then query with dig

```
Project/Twisted [ dig @localhost bing.com                             ] 7:57 PM

; <<>> DiG 9.7.3-P3 <<>> @localhost bing.com
; (3 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 30843
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;bing.com.			IN	A

;; ANSWER SECTION:
bing.com.		3600	IN	A	127.0.0.1

;; Query time: 1 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Sun Mar 11 19:57:23 2012
;; MSG SIZE  rcvd: 42
```

And the correct IP address is retrieved. However, if twisted is asked to recurse

```
sudo bin/twistd -n dns --hosts-file=hosts.txt -i 127.0.0.1 -r
```

And query for the same address

```
Project/Twisted [ dig @localhost bing.com                             ] 8:03 PM

; <<>> DiG 9.7.3-P3 <<>> @localhost bing.com
; (3 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 44932
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;bing.com.			IN	A

;; ANSWER SECTION:
bing.com.		179	IN	A	65.52.107.149

;; Query time: 98 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Sun Mar 11 20:05:20 2012
;; MSG SIZE  rcvd: 42
```

It appears that the recurse is being done before checking the hosts file. For unknown addresses (ones not really on the internet), everything appears to work correctly, since the recurse fails. The hosts file entry will then be used:

```
Project/Twisted [ dig @localhost bogus.domain.tld                     ] 8:14 PM

; <<>> DiG 9.7.3-P3 <<>> @localhost bogus.domain.tld
; (3 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 58695
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;bogus.domain.tld.		IN	A

;; ANSWER SECTION:
bogus.domain.tld.	3600	IN	A	127.0.0.1

;; Query time: 119 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Sun Mar 11 20:14:21 2012
;; MSG SIZE  rcvd: 50
```

Attachments:

* [dns-hosts-recursive-fix.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/176/17675597da5a132ec472b4ed560a9d9a92082959/c2f08820e7c643354a589215b09ce38557ab4513.patch) (672 bytes) - added by greezybacon on 2012-03-12 06:03:37Z - Server should check hosts file FIRST, then recurse
* [dns-hosts-recursive-fix.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/176/17675597da5a132ec472b4ed560a9d9a92082959/1dbe03697b597b806070d6228ee9e6eeb2ccb866.patch) (2936 bytes) - added by greezybacon on 2012-03-12 17:20:22Z - Better patch, with an associated  test case
* [dns-hosts-recursive-fix.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/176/17675597da5a132ec472b4ed560a9d9a92082959/4622cae150742161420e27aacb66996b644b2fbc.patch) (3032 bytes) - added by greezybacon on 2012-03-12 17:39:41Z - Even better patch, now only cleans up the parseConfig delayed call for the Resolver
* [dns-hosts-recursive-fix.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/176/17675597da5a132ec472b4ed560a9d9a92082959/4f78c68028c4acfa462842022fdc09a17a94551a.patch) (2819 bytes) - added by greezybacon on 2012-03-12 20:46:49Z - Move import statement back into the _buildConfig function

<details><summary>Searchable metadata</summary>

```
trac-id__5524 5524
type__defect defect
reporter__greezybacon greezybacon
priority__normal normal
milestone__ 
branch__branches_dns_recursion_5524_2 branches/dns-recursion-5524-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__dns_names_recurse dns names recurse
time__1331522588000000 1331522588000000
changetime__1332081871000000 1332081871000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="keturn's avatar" src="https://avatars.githubusercontent.com/u/83819?s=50" width="50" height="50">](https://github.com/keturn)|@keturn set owner to @keturn|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="keturn's avatar" src="https://avatars.githubusercontent.com/u/83819?s=50" width="50" height="50">](https://github.com/keturn)<a name="note_3"></a>|@keturn commented|
|-|-|

You've moved the import statement up to the module level, which makes sense in most cases, but tap files are an exception.  Reason being that all the tap files are loaded every time you run twistd, so you want to keep down the number of imports.

Oh, yow, you're still working on this.  Was the "review" keyword added prematurely?

**Comment by twisted-trac** [comment]
|[<img alt="keturn's avatar" src="https://avatars.githubusercontent.com/u/83819?s=50" width="50" height="50">](https://github.com/keturn)<a name="note_4"></a>|@keturn set owner to greezybacon|
|-|-|

The _buildResolvers refactoring looks like a good move for your testability, and the revision to the tests in the .3.patch is certainly a step in the right direction.

I'm trying to find someone to review the basic DNS policy question here, as it may have compatibility implications, and everyone I've asked about DNS so far says "talk to exarkun."

So I suggest fixing the imports, putting this back in review, and we'll see if we can get him to comment.

**Comment by twisted-trac** [comment]
|<img alt="greezybacon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|greezybacon commented|
|-|-|

RFC 1034 [http://www.ietf.org/rfc/rfc1034.txt], section 5.3.3 Step 1 indicates that information should be fetched from cache and local server's zone first, and then recurse the query to an appropriate server

I want to point out also, that the "recursive" resolver in this animal uses an instance of resolve.ResolverChain, which will actually have it own, separate instance of a cache resolver and a separate instance of a hosts resolver (which will use the system hosts file). In the ResolverChain, the local hosts file will be used in favor of recursing if the name is defined there.

The change this patch makes is, if the server is run with a separate --hosts-file option, then an _additional_ hosts file will be used, and should be used in favor of recursing to the resolve.ResolverChain resolver to recurse.

So the chain before the fix would be:

Cache -> Resolver { Cache -> System Hosts -> Recurse } -> Hosts { extra file provided }

I'm suggesting that this flow more like

Cache -> Hosts { extra file provided } -> Resolver { Cache -> System Hosts -> Recurse }

Honestly, the whole chain should be refactored more like

Cache -> Hosts { extra file provided } -> Hosts { system hosts file } -> Recurse (client.Resolver)

So that the second cache lookup is avoided

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Thanks!  This patch looks good to me now.  It needs a ReviewProcess#Newsfile, but I can easily add that.  I also agree with your remarks in comment 5 about further changes that should be made.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [33835]) Apply dns-hosts-recursive-fix.4.patch - fixing hosts/upstream precedence rules for "twistd dns"

Author: greezybacon
Reviewer: acapnotic, exarkun
Fixes: [#5524](https://github.com/twisted/twisted/issues/5524)

Change `twisted.names.tap` to construct a chain resolver that uses a hosts file resolver first,
and then falls back to a recursive resolver if necessary.  Previously a hosts file would not
override existing domain names that could be resolved over the network.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `reopened`|
|-|-|

(In [33880]) Revert r33835 - Windows test suite regression

Reopens: [#5524](https://github.com/twisted/twisted/issues/5524)

The new tests fail on Windows:

[ERROR]
Traceback (most recent call last):
  File "c:\bot-glyph-5\winxp32-py2.5-select\Twisted\twisted\names\test\test_tap.py", line 94, in test_recursiveConfiguration
    x.resolvers[-1]._parseCall.cancel()
  File "c:\bot-glyph-5\winxp32-py2.5-select\Twisted\twisted\names\root.py", line 431, in __getattr__
    raise AttributeError(name)
exceptions.AttributeError: _parseCall

twisted.names.test.test_tap.OptionsTests.test_recursiveConfiguration

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [33883]) Branching to 'dns-recursion-5524-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

I added a check to avoid the cleanup on Windows.  Let's see if that [works any better](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dns-recursion-5524-2).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [33925]) Correct api link; there is no usage.Config, and the real target should be the Options class in this module anyway.

refs [#5524](https://github.com/twisted/twisted/issues/5524)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [33926]) Merge dns-recursion-5524-2

Author: greezybacon, exarkun
Reviewer: acapnotic, exarkun
Fixes: [#5524](https://github.com/twisted/twisted/issues/5524)

Change `twisted.names.tap` to construct a chain resolver that uses a hosts file resolver first,
and then falls back to a recursive resolver if necessary.  Previously a hosts file would not
override existing domain names that could be resolved over the network.
